src/machines/java/com/enderio/machines/common/item/FluidTankItem.java

```diff
@@ -26,4 +50,62 @@ public BlockEntityWithoutLevelRenderer getCustomRenderer() {
             }
         });
     }
+
+    @Override
+    public void appendHoverText(ItemStack stack, @Nullable Level pLevel, List<Component> tooltip, TooltipFlag pFlag) {
+        ItemTooltip.addShiftKeyMessage(tooltip);
+        Optional<IFluidHandlerItem> fluidHandler = stack.getCapability(ForgeCapabilities.FLUID_HANDLER_ITEM).resolve();
+        if (fluidHandler.isPresent()){
+            ItemTooltip.addFluidTankMessage(tooltip, fluidHandler.get().getFluidInTank(0), capacity);//always present
+        } else {
+            EnderIO.LOGGER.warn("No fluid handler capability on tank item, unable to complete tooltip");
+        }
+
+        super.appendHoverText(stack, pLevel, tooltip, pFlag);
+    }
+    @Nullable
+    @Override
+    public ICapabilityProvider initCapabilities(ItemStack stack, @Nullable CompoundTag nbt) {
+        return new FluidItemStack(stack, capacity);
+    }
+
+    public class FluidItemStack extends FluidHandlerItemStack{
+        private static final String BLOCK_ENTITY_TAG = "BlockEntityTag";
+        /**
+         * @param container The container itemStack, data is stored on it directly as NBT.
+         * @param capacity  The maximum capacity of this fluid tank.
+         */
+        public FluidItemStack(@NotNull ItemStack container, int capacity) {
+            super(container, capacity);
+        }
+
+        @NotNull
+        public FluidStack getFluid() {
+            Optional<CompoundTag> tagCompoundOptional = Optional.ofNullable(container.getTag());
+            return tagCompoundOptional.map(tagCompound -> tagCompound.getCompound(BLOCK_ENTITY_TAG))
+                .map(beTag -> beTag.getCompound("fluid"))
+                .map(FluidStack::loadFluidStackFromNBT)
+                .orElse(FluidStack.EMPTY);
+        }
+
+        protected void setFluid(FluidStack fluid) {
+            CompoundTag mainTag = container.getOrCreateTag();
+            CompoundTag beTag = new CompoundTag();
+            mainTag.put(BLOCK_ENTITY_TAG, beTag);
+
+            CompoundTag fluidTag = new CompoundTag();
+            fluid.writeToNBT(fluidTag);//rewrites the old value
+            //TODO: externalize "fluid" into one global parameter.
```

Comment:
Could you clarify what you mean by this?

Reply:
Sure, since the tag name "fluid" is written to and read from in both the item variant and the block variant, as well as in many other parts of the mod. If you change it in one place, you are going to break the link between the two types. This bug would probably go unnoticed since it would only void fluids under very specific circumstances. So keeping the tag name in one, global place would greatly reduce this risk. It would simply be neater.